web/packages/teleport/src/components/hooks/useKeyBasedPagination.ts

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useState, useEffect, useRef } from 'react';
+import useAttempt, { Attempt } from 'shared/hooks/useAttemptNext';
+
+import {
+  ResourcesResponse,
+  ResourceFilter,
+  resourceFilterToHookDeps,
+} from 'teleport/services/agents';
+import { UrlResourcesParams } from 'teleport/config';
+
+/**
+ * Supports fetching more data from the server when more data is available. Pass
+ * a `fetchFunc` that retrieves a single batch of data. After the initial
+ * request, the server is expected to return a `startKey` field that denotes the
+ * next `startKey` to use for the next request.
+ *
+ * This hook is intended to be used in tandem with the `useInfiniteScroll` hook.
+ * See its documentation for an example.
+ */
+export function useKeyBasedPagination<T>({
+  fetchFunc,
+  clusterId,
+  filter,
+  initialFetchSize = 30,
+  fetchMoreSize = 20,
+}: Props<T>): State<T> {
+  const abortController = useRef<AbortController | null>(null);
+  const { attempt, setAttempt } = useAttempt();
+  const [finished, setFinished] = useState(false);
+  const [resources, setResources] = useState<T[]>([]);
+  const [startKey, setStartKey] = useState(null);
+
+  useEffect(() => {
+    abortController.current?.abort();
+    abortController.current = null;
+
+    setAttempt({ status: '', statusText: '' });
+    setFinished(false);
+    setResources([]);
+    setStartKey(null);
+  }, [clusterId, ...resourceFilterToHookDeps(filter)]);
+
+  // This is an additional countermeasure to prevent fetching if `fetch` is
+  // called multiple times per React state reconciliation cycle.
+  let fetchCalled = false;
+  const fetch = async (force: boolean) => {
+    if (
+      finished ||
+      fetchCalled ||
+      (!force &&
+        (attempt.status === 'processing' || attempt.status === 'failed'))
+    ) {
+      return;
+    }
+
+    try {
+      fetchCalled = true;
+      setAttempt({ status: 'processing' });
+      const limit = resources.length > 0 ? fetchMoreSize : initialFetchSize;
+      abortController.current?.abort();
+      abortController.current = new AbortController();
+      const res = await fetchFunc(
+        clusterId,
+        {
+          ...filter,
+          limit,
+          startKey,
+        },
+        abortController.current.signal
+      );
+      abortController.current = null;
```

Comment:
If we're setting this to `null` immediately after passing it's `signal`, how are we ever going to retrieve it again to `abort()`? I think this means that the `abortController.current?.abort();` line above always just returns `undefined`/`null`.

Reply:
Ha! That's because in line 83, we `await`, and by the time we reach this statement, `abort` on this controller wouldn't do anything at all, anyway. Technically, it shouldn't be absolutely necessary to set it to `null` here, but I wanted to maintain an invariant: if there's an abort controller, calling `abort` will actually abort something. Not sure why, I blame my OCD. :D

I'll add a clarifying comment.

Reply:
Ahh, I see, makes sense. Though in that case I prefer that you simplify by removing the unnecessary `null`ability.